File: src/year2022/day12/day12.rs
//! [Day 12: Hill Climbing Algorithm](https://adventofcode.com/2022/day/12)

use rustc_hash::FxHashSet;
use std::collections::VecDeque;

#[derive(Debug)]
struct Node {
    pos: (usize, usize),
    steps: u32,
}

struct Puzzle {
    grid: aoc::GridU<u8>,
    start: (usize, usize),
    end: (usize, usize),
}

impl Puzzle {
    fn new(data: &str) -> Self {
        let mut puzzle = Self {
            grid: aoc::GridU::<u8>::parse(data.trim_ascii()),
            start: (0, 0),
            end: (0, 0),
        };

        for (pos, c) in puzzle.grid.iter_mut() {
            *c = match c {
                b'S' => {
                    puzzle.start = pos; // start position: height 1
                    1
                }
                b'E' => {
                    puzzle.end = pos; // end position: height 26 (like 'z')}
                    26
                }
                b'a'..=b'z' => *c - b'a' + 1, // other locations: height 1 to 26
                _ => panic!("unknown elevation {c}"),
            };
        }

        puzzle
    }

    /// Breadth-first search
    fn bfs(&self, part: u8) -> u32 {
        let mut q: VecDeque<Node> = VecDeque::new();
        let mut seen: FxHashSet<(usize, usize)> = FxHashSet::default();

        // set up the start positions:
        // 'S' if part 1
        // 'S' or any 'a' if part 2 (height 1)
        if part == 1 {
            q.push_back(Node {
                pos: self.start,
                steps: 0,
            });
        } else {
            for (pos, c) in self.grid.iter() {
                if *c == 1 {
                    let n = Node { pos, steps: 0 };
                    q.push_back(n);
                }
            }
        }

        while !q.is_empty() {
            let n = q.pop_front().unwrap();
            let pos = n.pos;

            if pos == self.end {
                return n.steps;
            }

            if !seen.insert(pos) {
                continue;
            }

            for new_pos in self.grid.iter_directions(pos) {
                if self.grid[new_pos] <= 1 + self.grid[pos] {
                    q.push_back(Node {
                        pos: new_pos,
                        steps: n.steps + 1,
                    });
                }
            }
        }

        0
    }

    // Solves part one
    fn part1(&self) -> u32 {
        self.bfs(1)
    }

    // Solve part two
    fn part2(&self) -> u32 {
        self.bfs(2)
    }
}

#[must_use]
pub fn solve(data: &str) -> (u32, u32) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 31);
        assert_eq!(puzzle.part2(), 29);
    }
}

File: src/year2022/day12/day12.py
#!/usr/bin/env python3
# [Day 12: Hill Climbing Algorithm](https://adventofcode.com/2022/day/12)

import atexit
import sys
import time
from collections import deque
from pathlib import Path

filename = (
    "test.txt" if len(sys.argv) > 1 and sys.argv[1] == "-t" else sys.argv[1] if len(sys.argv) > 1 else "input.txt"
)
data = Path(filename).read_text().strip()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


grid = []
for y, line in enumerate(data.splitlines()):
    row = []
    for x, c in enumerate(line):
        if c == "E":
            end_position = (x, y)
            c = 26
        elif c == "S":
            start_position = (x, y)
            c = 1
        else:
            c = ord(c) - 96
        row.append(c)
    grid.append(row)
nx, ny = len(grid[0]), len(grid)


def bfs(part):
    """[Breadth-first search](https://en.wikipedia.org/wiki/Breadth-first_search)"""

    q = deque()
    if part == 1:
        q.append((start_position, 0))
    else:
        for y in range(ny):
            for x in range(nx):
                if grid[y][x] == 1:  # start at 'S' or any 'a'
                    q.append(((x, y), 0))

    visited = set()
    while q:
        (x, y), steps = q.popleft()
        if (x, y) in visited:
            continue
        visited.add((x, y))
        if (x, y) == end_position:
            return steps
        for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
            x2, y2 = x + dx, y + dy
            # move within the grid and « at most one higher »
            if 0 <= x2 < nx and 0 <= y2 < ny and grid[y2][x2] <= 1 + grid[y][x]:
                q.append(((x2, y2), steps + 1))


print(bfs(1))
print(bfs(2))
